File: app/src/main/java/com/mirash/familiar/activity/pin/PinActivityModel.java
package com.mirash.familiar.activity.pin;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.mirash.familiar.Const;
import com.mirash.familiar.view.PinInputView;

import java.util.Stack;

/**
 * @author Mirash
 */
public class PinActivityModel extends AndroidViewModel implements PinInputView.PinCallback {
    private final MutableLiveData<Integer> pinCodeSizeLiveData;
    private final MutableLiveData<String> pinCodeCheckLiveData;
    private final Stack<Integer> pinCode;

    public PinActivityModel(@NonNull Application application) {
        super(application);
        pinCode = new Stack<>();
        pinCodeSizeLiveData = new MutableLiveData<>(0);
        pinCodeCheckLiveData = new MutableLiveData<>(null);
    }

    @Override
    public void onDigitClick(int digit) {
        if (pinCode.size() >= Const.PIN_CODE_SIZE) return;
        pinCode.push(digit);
        int size = pinCode.size();
        pinCodeSizeLiveData.setValue(size);
        if (size == Const.PIN_CODE_SIZE) {
            checkPinCode();
        }
    }

    @Override
    public void onBackspaceClick() {
        if (pinCode.isEmpty()) return;
        pinCode.pop();
        pinCodeSizeLiveData.setValue(pinCode.size());
    }

    public MutableLiveData<Integer> getPinCodeSizeLiveData() {
        return pinCodeSizeLiveData;
    }

    public MutableLiveData<String> getPinCodeCheckLiveData() {
        return pinCodeCheckLiveData;
    }

    private void checkPinCode() {
        StringBuilder builder = new StringBuilder();
        for (Integer digit : pinCode) {
            builder.append(digit);
        }
        String pinCode = builder.toString();
        pinCodeCheckLiveData.setValue(pinCode);
    }
}

File: app/src/main/java/com/mirash/familiar/activity/pin/PinActivityModel.kt
package com.mirash.familiar.activity.pin

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.mirash.familiar.tool.PIN_CODE_SIZE
import com.mirash.familiar.view.PinInputView
import java.util.Stack

/**
 * @author Mirash
 */
class PinActivityModel(application: Application) : AndroidViewModel(application), PinInputView.PinCallback {
    val pinCodeSizeLiveData: MutableLiveData<Int> = MutableLiveData<Int>(0)
    val pinCodeCheckLiveData: MutableLiveData<String> = MutableLiveData<String>(null)
    private val pinCode: Stack<Int> = Stack()

    override fun onDigitClick(digit: Int) {
        if (pinCode.size >= PIN_CODE_SIZE) return
        pinCode.push(digit)
        val size = pinCode.size
        pinCodeSizeLiveData.value = size
        if (size == PIN_CODE_SIZE) {
            checkPinCode()
        }
    }

    override fun onBackspaceClick() {
        if (pinCode.isEmpty()) return
        pinCode.pop()
        pinCodeSizeLiveData.value = pinCode.size
    }

    private fun checkPinCode() {
        val builder = StringBuilder()
        for (digit in pinCode) {
            builder.append(digit)
        }
        val pinCode = builder.toString()
        pinCodeCheckLiveData.value = pinCode
    }
}
